### ISSUE TITLE
ActiveRecord fails to include models given on a particular order after a join

### ISSUE THREAD

**Issue by clemthenem** [rails/rails#30603]
#### Setup
I have three models
``` ruby
# models/profile_translation.rb
class ProfileTranslation < ApplicationRecord
  belongs_to :profile
  has_one :user, through: :profile
end

# models/profile.rb
class Profile < ApplicationRecord
  belongs_to :user
  has_many :profile_translations, foreign_key: 'profile_id'
end

# models/user.rb
class User < ApplicationRecord
  has_one :profile
  has_many :profile_translations, through: :profile
end
```

#### Query
This query (which looks silly but seems valid to me) is to simplify a much bigger example.
``` ruby
ProfileTranslation.joins(:profile).includes(:user, :profile)
```

### Expected behavior
I would expect to get an ActiveRecord::Relation of ProfileTranslations along with their user and profile.

### Actual behavior
``` ruby
PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "profiles_profile_translations"...
```

The interesting part is that if I change the order or the included elements:
``` ruby
ProfileTranslation.joins(:profile).includes(:profile, :user)
```
Then I get my expected result.

### System configuration
**Rails version**:
gem 'rails', '~> 5.1.4'

**Ruby version**:
ruby '2.4.1'

**Comment by kamipo** [comment]
Can you provide an executable test script using [this template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb)?

**Comment by Bartuz** [comment]
Couldn't recreate the issue. Here is my executable test script, @kamipo ,

https://gist.github.com/Bartuz/175845efab2ab3f3bc6b808cced7f571

@clemthenem , are you sure it isn't something about your applicationRecord?

**Comment by houhoulis** [comment]
@Bartuz , the reason your script did not replicate the issue is because your code creates a query but doesn't execute it.  If you append a ruby command like "`map(&:id)`" to the end of your query, then it does reproduce the error.

**Comment by houhoulis** [comment]
Using ruby 2.4.2, I was able to replicate the behavior that @clemthenem described by using @Bartuz 's script with "`.map(&:class)`" appended to the query.  I observed different error messages when using the 'pg' gem vs. the 'sqlite3' gem: _"PG::UndefinedTable: ERROR:  missing FROM-clause entry for table 'profiles_profile_translations'"_ vs _"SQLite3::SQLException: no such column: profiles_profile_translations.id"_.  But both cases logged the same generated SQL to stdout:

```ruby
ProfileTranslation.joins(:profile).includes(:user, :profile).map(&:class)
  SELECT "profile_translations"."id" AS t0_r0, "profile_translations"."profile_id" AS t0_r1,
  "users"."id" AS t1_r0, "profiles_profile_translations"."id" AS t2_r0,
  "profiles_profile_translations"."user_id" AS t2_r1
  FROM "profile_translations"
  INNER JOIN "profiles" ON "profiles"."id" = "profile_translations"."profile_id"
  LEFT OUTER JOIN "profiles" "profiles_profile_translations_join"
    ON "profiles_profile_translations_join"."id" = "profile_translations"."profile_id"
  LEFT OUTER JOIN "users"
    ON "users"."id" = "profiles_profile_translations_join"."user_id"
```

And the contrasting example did produce SQL without the phantom table:

```ruby
ProfileTranslation.joins(:profile).includes(:profile, :user).map(&:class)
  SELECT "profile_translations"."id" AS t0_r0, "profile_translations"."profile_id" AS t0_r1,
  "profiles"."id" AS t1_r0, "profiles"."user_id" AS t1_r1, "users"."id" AS t2_r0
  FROM "profile_translations"
  INNER JOIN "profiles" ON "profiles"."id" = "profile_translations"."profile_id"
  LEFT OUTER JOIN "profiles" "profiles_profile_translations_join"
    ON "profiles_profile_translations_join"."id" = "profile_translations"."profile_id"
  LEFT OUTER JOIN "users"
    ON "users"."id" = "profiles_profile_translations_join"."user_id"
```

I don't know enough to evaluate @clemthenem 's symptoms, nor whether the example is well-formed.

**Comment by clemthenem** [comment]
Hi guys,
Thanks a lot @Bartuz and @houhoulis for your time replicating the issue, I apologise for not replying earlier... The example is well formed yes, that's exactly what happened to me. Let me know if you need any extra info @kamipo and thank you so much for your help!

**Comment by ziaulrehman40** [comment]
Was this fix released in rails 4 line or only 5.1.x, i am sorry not sure how to check that.

**Comment by kamipo** [comment]
This is fixed in master, and backported to 5-2-stable in 50036e673b3b6ff41476814b212b55fabac3a638.
